src/c++/perf_analyzer/infer_data_manager_factory.h

```diff
@@ -36,14 +36,15 @@ namespace triton { namespace perfanalyzer {
 class InferDataManagerFactory {
  public:
   static std::shared_ptr<IInferDataManager> CreateInferDataManager(
-      const int32_t batch_size, const SharedMemoryType shared_memory_type,
-      const size_t output_shm_size, const std::shared_ptr<ModelParser>& parser,
+      const size_t max_threads, const int32_t batch_size,
+      const SharedMemoryType shared_memory_type, const size_t output_shm_size,
+      const std::shared_ptr<ModelParser>& parser,
       const std::shared_ptr<cb::ClientBackendFactory>& factory,
       const std::shared_ptr<DataLoader>& data_loader)
   {
     if (shared_memory_type == SharedMemoryType::NO_SHARED_MEMORY) {
       return CreateInferDataManagerNoShm(
-          batch_size, parser, factory, data_loader);
+          max_threads, batch_size, parser, factory, data_loader);
```

Comment:
High level question: why is max threads and thread_id not needed by shm version?

Reply:
Good question. The InferData objects that we are precreating for the non-shm case are not thread safe. They implement internal iterators, and if multiple threads are "getting" data from them at the same time it won't work correctly. As a result, I had to create copies for every thread.

For the SHM case, the precreated input data is put into shared memory instead of in InferData objects. Then, on each request, the InferData object that the thread already owns is pointed to the shared memory.

It is safe to share shared memory, but it is not safe to share InferData objects.